internal/controller/commitstatus_controller.go

```diff
@@ -65,36 +66,56 @@ func (r *CommitStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	err := r.Get(ctx, req.NamespacedName, &cs, &client.GetOptions{})
 	if err != nil {
 		if errors.IsNotFound(err) {
-			logger.Info("CommitStatus not found", "namespace", req.Namespace, "name", req.Name)
+			logger.Info("CommitStatus not found")
 			return ctrl.Result{}, nil
 		}
 
-		logger.Error(err, "failed to get CommitStatus", "namespace", req.Namespace, "name", req.Name)
+		logger.Error(err, "failed to get CommitStatus")
 		return ctrl.Result{}, err
 	}
 
 	// We use observed generation pattern here to avoid provider API calls.
 	if cs.Status.ObservedGeneration == cs.Generation {
-		logger.Info("No need to reconcile", "namespace", req.Namespace, "name", req.Name)
+		logger.Info("No need to reconcile")
 		return ctrl.Result{}, nil
 	}
 
-	commitStatusProvider, err := r.getCommitStatusProvider(ctx, cs)
-	if err != nil {
-		return ctrl.Result{}, err
+	if cs.Spec.Sha == "" || cs.Spec.Phase == "" {
+		logger.Info("Skip setting commit status, missing sha or phase")
+		return ctrl.Result{}, nil
 	}
 
-	commitStatus, err := commitStatusProvider.Set(ctx, &cs)
+	commitStatusProvider, err := r.getCommitStatusProvider(ctx, cs)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	r.Recorder.Eventf(&cs, "Normal", "CommitStatusSet", "Commit status %s set to %s", commitStatus.Name, commitStatus.Spec.Phase)
 
-	commitStatus.Status.ObservedGeneration = commitStatus.Generation
-	err = r.Status().Update(ctx, commitStatus)
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		newCs := promoterv1alpha1.CommitStatus{}
+		err := r.Client.Get(ctx, req.NamespacedName, &newCs, &client.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		_, err = commitStatusProvider.Set(ctx, &cs)
+		if err != nil {
+			return err
+		}
+
+		newCs.Status.ObservedGeneration = newCs.Generation
```

Comment:
Not sure I follow why `newCs` exists, it doesn't look like we're actually using it.

Reply:
newCs gets updated from the Get request

Reply:
Right, but then as far as I can tell it never gets used beyond this field mutation.

Reply:
Your right that should be used in the update